### PR TITLE
feat: capture the environment where the connection was created with

### PR DESCRIPTION
**Pull request by yutannihilation**
My attempt for https://github.com/yutannihilation/savvy/pull/254.

Note that this implementation is probably incorrect. I manually mark `Send` and `Sync`, but we cannot call R API concurrently anyway.

### CODE REVIEW

**Comment by eitsupi** [comment]
Thanks for working on this!

This implementation would search for objects in the environment where the connection was created.
(It is probably better to fix the default connection with Global env, since it is initialized on the Rust side.)

```r
out <- glaredb_sql("SELECT 'foo' as x")

fn1 <- function() {
  con <- glaredb_connect()
  out <- glaredb_sql("SELECT 'bar' as y")
  con$sql("select * from out") |> as.data.frame()
}

fn2 <- function() {
  con <- glaredb_connect(env = parent.frame())
  out <- glaredb_sql("SELECT 'bar' as y")
  con$sql("select * from out") |> as.data.frame()
}

fn1()
#>     y
#> 1 bar
fn2()
#>     x
#> 1 foo
```

I think it is much better than before, considering its intended use.

**Comment by eitsupi** [comment]
Is it possible to use loop to move to the parent environment and explore recursively, as duckdb is doing here?
https://github.com/duckdb/duckdb-r/blob/d18690cd642ceca238b54e43fbeffd6f340e45d4/src/register.cpp#L58-L64

If we could do this it would work perfectly.

**Comment by eitsupi** [comment]
It seems to be working, so I will merge this. Thanks for your great work!

File: src/rust/src/environment.rs [hunk 1]
```diff
@@ -3,31 +3,60 @@ use arrow::array::RecordBatchReader;
...
+            token: UnsafeToken(token),
+        }
+    }
+}
+
+impl Drop for REnvironmentReader {
+    fn drop(&mut self) {
+        release_from_preserved_list(self.token.0)
+    }
+}
```
Comment (eitsupi) [inline src/rust/src/environment.rs:39]: @yutannihilation Sorry to bother you.
This package is AGPL-3.0 licensed, but may I copy this to https://github.com/pola-rs/r-polars and distribute it under the terms of the MIT license (re-licensing)?

Comment (yutannihilation) [reply src/rust/src/environment.rs:39]: Yes, please feel free to use with any license! (I'm not fully confident if this is the best option, though)

Comment (eitsupi) [reply src/rust/src/environment.rs:39]: Thanks!
Thanks for your help, R udf worked.
https://github.com/eitsupi/neo-r-polars/commit/a9893624feb6087bca8412cedc8558a67c1af48a

Comment (yutannihilation) [reply src/rust/src/environment.rs:39]: Great progress!

**Pull request merged**